pkg/applicationserver/io/web/webhooks.go

```diff
@@ -423,39 +335,49 @@ func (w *webhooks) newRequest(ctx context.Context, msg *ttnpb.ApplicationUp, hoo
 
 var errWebhookNotFound = errors.DefineNotFound("webhook_not_found", "webhook not found")
 
-func (w *webhooks) handleDown(c echo.Context, op func(io.Server, context.Context, ttnpb.EndDeviceIdentifiers, []*ttnpb.ApplicationDownlink) error) error {
-	ctx := w.server.FillContext(c.Request().Context())
-	devID := c.Get(deviceIDKey).(ttnpb.EndDeviceIdentifiers)
-	hookID := c.Get(webhookIDKey).(ttnpb.ApplicationWebhookIdentifiers)
-	logger := log.FromContext(ctx).WithFields(log.Fields(
-		"application_id", devID.ApplicationID,
-		"device_id", devID.DeviceID,
-		"webhook_id", hookID.WebhookID,
-	))
-	hook, err := w.registry.Get(ctx, hookID, []string{"format"})
-	if err != nil {
-		return err
-	}
-	if hook == nil {
-		return errWebhookNotFound.New()
-	}
-	format, ok := formats[hook.Format]
-	if !ok {
-		return errFormatNotFound.WithAttributes("format", hook.Format)
-	}
-	body, err := ioutil.ReadAll(c.Request().Body)
-	if err != nil {
-		return err
-	}
-	items, err := format.ToDownlinks(body)
-	if err != nil {
-		return err
-	}
-	logger.Debug("Perform downlink queue operation")
-	if err := op(w.server, ctx, devID, items.Downlinks); err != nil {
-		return err
-	}
-	return nil
+func (w *webhooks) handleDown(op func(io.Server, context.Context, ttnpb.EndDeviceIdentifiers, []*ttnpb.ApplicationDownlink) error) http.Handler {
+	return http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		devID := ctx.Value(deviceIDKey).(ttnpb.EndDeviceIdentifiers)
+		hookID := ctx.Value(webhookIDKey).(ttnpb.ApplicationWebhookIdentifiers)
```

Comment:
Maybe better to create funcs for these in `pkg/applicationserver/io/web/middleware.go` so that insert/extract into/from context is in the same place. That would also make it easier to find out what's wrong if the middleware would be removed (by accident) and this would panic without explanation.